### PR TITLE
[SPARK-11563] [core] [repl] Use RpcEnv to transfer REPL-generated classes.

### PR DESCRIPTION
**Pull request by vanzin**
This avoids bringing up yet another HTTP server on the driver, and
instead reuses the file server already managed by the driver's
RpcEnv. As a bonus, the repl now inherits the security features of
the network library.

There's also a small change to create the directory for storing classes
under the root temp dir for the application (instead of directly
under java.io.tmpdir).

### CODE REVIEW

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
retest this please

File: repl/src/test/scala/org/apache/spark/repl/ExecutorClassLoaderSuite.scala [hunk 1]
```diff
@@ -18,24 +18,29 @@
...
 import org.scalatest.concurrent.Interruptor
 import org.scalatest.concurrent.Timeouts._
 import org.scalatest.mock.MockitoSugar
+import org.mockito.invocation.InvocationOnMock
+import org.mockito.stubbing.Answer
+import org.mockito.Matchers.anyString
 import org.mockito.Mockito._
 
 import org.apache.spark._
+import org.apache.spark.rpc.{RpcEnv, RpcEnvFileServer}
```
Comment (squito) [inline repl/src/test/scala/org/apache/spark/repl/ExecutorClassLoaderSuite.scala]: super nit: RpcEnvFileServer is unused

File: repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkIMain.scala [hunk 2]
```diff
@@ -96,10 +96,9 @@ import org.apache.spark.annotation.DeveloperApi
 
     private val SPARK_DEBUG_REPL: Boolean = (System.getenv("SPARK_DEBUG_REPL") == "1")
     /** Local directory to save .class files too */
-    private lazy val outputDir = {
-      val tmp = System.getProperty("java.io.tmpdir")
-      val rootDir = conf.get("spark.repl.classdir",  tmp)
-      Utils.createTempDir(rootDir)
+    private[repl] val outputDir = {
+      val rootDir = conf.getOption("spark.repl.classdir").getOrElse(Utils.getLocalDir(conf))
```
Comment (squito) [inline repl/scala-2.10/src/main/scala/org/apache/spark/repl/SparkIMain.scala]: this looks like the right change, but probably worth mentioning in the PR description that you're also changing the default dir where the repl class files get written.

**Comment by squito** [comment]
looks reasonable to me, I tested out a repl with both netty & akka in local-cluster mode as well.  Would appreciate another set of eyes on it.

**Comment by vanzin** [comment]
Given the silence I guess people are ok with the patch, so I'll push this soon. Meanwhile, retest this please.

File: repl/src/main/scala/org/apache/spark/repl/ExecutorClassLoader.scala [hunk 3]
```diff
@@ -90,6 +94,11 @@ class ExecutorClassLoader(conf: SparkConf, classUri: String, parent: ClassLoader
     }
   }
 
+  private def getClassFileInputStreamFromSpark(path: String): InputStream = {
```
Comment (rxin) [inline repl/src/main/scala/org/apache/spark/repl/ExecutorClassLoader.scala]: getClassFileInputStreamFromSpark -> getClassFileInputStreamFromSparkRPC

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala [hunk 4]
```diff
@@ -41,7 +42,10 @@ private[netty] class NettyStreamManager(rpcEnv: NettyRpcEnv)
     val file = ftype match {
       case "files" => files.get(fname)
       case "jars" => jars.get(fname)
-      case _ => throw new IllegalArgumentException(s"Invalid file type: $ftype")
+      case  other =>
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala]: an extra space here?

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala [hunk 5]
```diff
@@ -41,7 +42,10 @@ private[netty] class NettyStreamManager(rpcEnv: NettyRpcEnv)
...
+        val dir = dirs.get(ftype)
+        require(dir != null, s"Invalid stream URI: $ftype not found.")
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala]: where does this exception propagate to?

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala [hunk 6]
```diff
@@ -31,6 +31,7 @@ private[netty] class NettyStreamManager(rpcEnv: NettyRpcEnv)
 
   private val files = new ConcurrentHashMap[String, File]()
   private val jars = new ConcurrentHashMap[String, File]()
+  private val dirs = new ConcurrentHashMap[String, File]()
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala]: while you are at it, can you update the classdoc for this to explain the different things this manager can serve, and explain the difference between jars and files?

File: core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala [hunk 7]
```diff
@@ -179,6 +179,24 @@ private[spark] trait RpcEnvFileServer {
...
+   *
+   * @param baseUri Leading URI path (files can be retrieved by appending their relative
+   *                path to this base URI). This cannot be "files" nor "jars".
+   * @param path Path to the local directory.
+   * @return URI for the root of the directory in the file server.
+   */
+  def addDirectory(baseUri: String, path: File): String
+
+  /** Validates and normalizes the base URI for directories. */
+  protected def validateDirectoryUri(baseUri: String): String = {
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/rpc/RpcEnv.scala]: would be better to have a unit test for this

**Comment by rxin** [comment]
IIUC, http server is useless once we remove akka? Can you add a line of comment there as well?

LGTM.

File: core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala (hunk 5 above)
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/rpc/netty/NettyStreamManager.scala]: It's handled by `TransportRequestHandler.processStreamRequest`.

**Comment by vanzin** [comment]
> IIUC, http server is useless once we remove akka?

There's the http-based broadcast server, which IIRC is not the default. Does anyone even use that these days?

**Comment by vanzin** [comment]
Merging to master.

**Pull request closed without merging**